drivers/sensor/lsm6dsr/lsm6dsr_trigger.c

```diff
@@ -0,0 +1,312 @@
+ */
+static void lsm6dsr_handle_interrupt(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm6dsr_status_reg_t status;
+
+	while (1) {
+		if (lsm6dsr_status_reg_get(ctx, &status) < 0) {
+			LOG_DBG("failed reading status reg");
+			return;
+		}
+
+		if ((status.xlda == 0) && (status.gda == 0)
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+			&& (status.tda == 0)
+#endif
+		) {
+			break;
+		}
+
+		if ((status.xlda) && (lsm6dsr->handler_drdy_acc != NULL)) {
+			lsm6dsr->handler_drdy_acc(dev, lsm6dsr->trig_drdy_acc);
+		}
+
+		if ((status.gda) && (lsm6dsr->handler_drdy_gyr != NULL)) {
+			lsm6dsr->handler_drdy_gyr(dev, lsm6dsr->trig_drdy_gyr);
+		}
+
+		if (IS_ENABLED(CONFIG_LSM6DSR_ENABLE_TEMP)) {
+			if ((status.tda) && (lsm6dsr->handler_drdy_temp != NULL)) {
+				lsm6dsr->handler_drdy_temp(dev, lsm6dsr->trig_drdy_temp);
+			}
+		}
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy,
+					GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static void handle_irq(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy, GPIO_INT_DISABLE);
+
+	if (IS_ENABLED(CONFIG_LSM6DSR_TRIGGER_OWN_THREAD)) {
+		k_sem_give(&lsm6dsr->gpio_sem);
+	} else if (IS_ENABLED(CONFIG_LSM6DSR_TRIGGER_GLOBAL_THREAD)) {
+		k_work_submit(&lsm6dsr->work);
+	}
+}
+
+static void lsm6dsr_gpio_callback(const struct device *dev,
+				    struct gpio_callback *cb, uint32_t pins)
+{
+	struct lsm6dsr_data *lsm6dsr =
+		CONTAINER_OF(cb, struct lsm6dsr_data, gpio_cb);
+
+	ARG_UNUSED(pins);
+
+	handle_irq(lsm6dsr->dev);
+}
+
+static void lsm6dsr_thread(struct lsm6dsr_data *lsm6dsr)
+{
+	while (1) {
+		k_sem_take(&lsm6dsr->gpio_sem, K_FOREVER);
+		lsm6dsr_handle_interrupt(lsm6dsr->dev);
+	}
+}
+
+static void lsm6dsr_work_cb(struct k_work *work)
+{
+	struct lsm6dsr_data *lsm6dsr =
+		CONTAINER_OF(work, struct lsm6dsr_data, work);
+
+	lsm6dsr_handle_interrupt(lsm6dsr->dev);
+}
+
+int lsm6dsr_init_interrupt(const struct device *dev)
+{
+	struct lsm6dsr_data *lsm6dsr = dev->data;
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	int ret;
+
+	/* setup data ready gpio interrupt (INT1 or INT2) */
+	if (!device_is_ready(cfg->gpio_drdy.port)) {
+		LOG_ERR("Cannot get pointer to drdy_gpio device");
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_LSM6DSR_TRIGGER_OWN_THREAD)) {
+		k_sem_init(&lsm6dsr->gpio_sem, 0, K_SEM_MAX_LIMIT);
+
+		k_thread_create(&lsm6dsr->thread, lsm6dsr->thread_stack,
+				CONFIG_LSM6DSR_THREAD_STACK_SIZE,
+				(k_thread_entry_t)lsm6dsr_thread, lsm6dsr,
+				NULL, NULL, K_PRIO_COOP(CONFIG_LSM6DSR_THREAD_PRIORITY),
+				0, K_NO_WAIT);
+		k_thread_name_set(&lsm6dsr->thread, "lsm6dsr");
+	} else if (IS_ENABLED(CONFIG_LSM6DSR_TRIGGER_GLOBAL_THREAD)) {
+		lsm6dsr->work.handler = lsm6dsr_work_cb;
+	}
```

Comment:
This could be externalized into `lsm6dsr_trigger_init()` inside of `lsm6dsr_trigger.c`